pkg/tsdb/cloudwatch/cloudwatch.go

```diff
@@ -50,15 +53,21 @@ const logStreamIdentifierInternal = "__logstream__grafana_internal__"
 var plog = log.New("tsdb.cloudwatch")
 var aliasFormat = regexp.MustCompile(`\{\{\s*(.+?)\s*\}\}`)
 
+const defaultConcurrentQueries = 4
+
 func init() {
+	globalExecutor := newExecutor()
 	tsdb.RegisterTsdbQueryEndpoint("cloudwatch", func(ds *models.DataSource) (tsdb.TsdbQueryEndpoint, error) {
-		return newExecutor(), nil
+		return globalExecutor, nil
```

Comment:
What's the benefit of this change? Why is it necessary for this PR?

Reply:
@kaydelaney Thanks, have you considered if it could break any of the existing functionality though?

Reply:
The only potential issue I can think of at the moment involves how the cloudwatch logs clients are cached at the moment. Currently there's no invalidation mechanism so if some changes were made to the datasource by a user (default region, credentials, etc.) those changes wouldn't take effect unless the backend is restarted, which obviously isn't ideal. This could be fixed by just getting rid of the client caching though.

Reply:
I think this one could be quite a big risk. Have a feeling that cloudwatch datasource have stored temporary data on cloudwatch executor which have been used further down the call chain. Last time this happened was recently when we fixed a bug for custom metrics. Introducing this could mean that it has to be rewritten to the "thread safe"

I obviously don't have all the details and understanding. My focus (and Arve's) is to be able to merge this for 7.3.

Thinking out loud a bit here, but would like to get some better understandings so please let me know if I'm completely off.

Feels like changing to a "globalExecutor" was made because you need to have one instance of `queuesByRegion map[string](chan bool)` which I guess is the queue used for "handling" logs queries. To me it does look a bit strange  to only have one queue by region, but I don't know all the details. What happens if you have multiple cloudwatch data sources targeting different accounts - is there a risk pushing out logs from wrong account to a user?

My solution to not having a "globalExecutor" would be to define a new struct or interface that can be provided as dependency when creating a new `cloudWatchExecutor`. Example code:

```go
func init() {
  myLogQueueService := newLogQueueService()
  
  tsdb.RegisterTsdbQueryEndpoint("cloudwatch", func(ds *models.DataSource) (tsdb.TsdbQueryEndpoint, error) {
    return newExecutor(myLogQueueService), nil
  }
}
```

To take this a step further we can define a new struct/interface than handles everything related to a "liveLogAction". Maybe you share a lot of code between "logAction" and "liveLogAction" and this would not be optimal, but maybe there's some kind of middle way abstraction between "logQueue" and "liveLogActionHandler". In addition we could maybe ask the live package from cloudwatch to register the handler instead of the opposite order like now. Example code :

```go
func init() {
  myliveLogActionExecutor := newLiveLogActionExecutor()
  // register live handler
  live.RegisterPluginHandler("cloudwatch", myliveLogActionExecutor)
  tsdb.RegisterTsdbQueryEndpoint("cloudwatch", func(ds *models.DataSource) (tsdb.TsdbQueryEndpoint, error) {
    return newExecutor(myliveLogActionExecutor), nil
  }
}

func (e *cloudWatchExecutor) Query(ctx context.Context, dsInfo *models.DataSource, queryContext *tsdb.TsdbQuery) (*tsdb.Response, error) {
  ...
  case "liveLogAction":
    return e.liveLogActionExecutor.Execute(ctx, queryContext)
  ...
}
```

Going this path the liveLogAction handling will be completely separated from existing cloudwatch implementation and by that reduces risk and it also simplifies things according to me. Obviously more work to get in these changes, but I think they're needed to get this into 7.3.

@kaydelaney @aknuds1 I suggest that we meet tomorrow, if needed, and sync on/discuss the path forward for getting this merged. WDYT?

Reply:
I'm up for a meet today :)

Reply:
> Have a feeling that cloudwatch datasource have stored temporary data on cloudwatch executor which have been used further down the call chain

Spent some more time looking into this and while I think it should be fine for the cloudwatch logs side of things, I hadn't considered the metrics side of things and struct properties like `ec2Client` which get set and changed in a way that wouldn't be threadsafe with a global executor.

> Feels like changing to a "globalExecutor" was made because you need to have one instance of `queuesByRegion map[string](chan bool)` which I guess is the queue used for "handling" logs queries. To me it does look a bit strange to only have one queue by region, but I don't know all the details. What happens if you have multiple cloudwatch data sources targeting different accounts - is there a risk pushing out logs from wrong account to a user?

Re: Having multiple cloudwatch data sources targeting different accounts and how this might affect things, I had actually made a note for myself to investigate that but promptly forgot about it, so thank you for bringing it up, as it's a very good point. The queues would have to be by region and by account, so I suppose we could still have a map with the channels as before, but rather than just using the region as string, something like
`queueKey :=  region + "-" + accountID` would work better. To clarify though, as it is now it wouldn't result in the wrong logs getting pushed to users, it just means requests wouldn't be dispatched as quickly as they could be.

Reply:
Thanks :+1: It didn't really go as planned in regards to meet today, but @aknuds1 and me have started to look into a couple of tweaks of how we can improve. Lets sync up on Monday.